### DISCUSSION TITLE
ODFs given by Fourier coefficients and discreteSample uses issue

### DISCUSSION THREAD

**Discussion by moctarkaba** [mtex-toolbox/mtex#2144] [Ask Anything] [answered]
Hi ,
Firstly, I would like to say thank you for this great toolbox that is Mtex!! 
As of now, I'm working on texture analysis for hexagonal material with the VPSC code.

Due to that I was given the fourrier coefficient that will help me build the ODF with the **FourierODF** tool thanks to mtex.
Then extract the orientation points for VPSC.
However when I try to extract some orientation points with the tool **discreteSample**  for VPSC, I have some issues.  
Added to that, it seems that when I try to plot the pole figure I have a warning saying "the Imaginary part of complex valued SO3FunHarmonic is ignored". 
 For reference, here is the code lines concerned : 
 
CODE \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
cs = crystalSymmetry('6');
C = [C_0;C_1;C_2;C_3;C_4;C_5;C_6;C_8;C_9;C_10;C_11;C_12;C_13;C_14;C_15;C_16]; % Fourier coefficients (they are complex coefficient)
odf = SO3FunHarmonic(C,cs);
%odf = FourierODF(C,cs);
plotPDF(odf,[Miller(0,0,0,2,cs)],'antipodal')
mtexColorbar
ori = discreteSample(odf,3000);

OUTPUT      \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

Warning: Imaginary part of complex valued SO3FunHarmonic is ignored.
In the following only the real part is plotted. 
> In SO3FunHarmonic/plot (line 32)
In coeff_fourier (line 277) 
Error using histcounts>validateBinEdges
'BinEdges' must be a non-empty, real, numeric vector with no missing
values.

Error in histcounts (line 123)
        validateBinEdges(a);

Error in discretesample (line 108)
c = histcounts(rv, edges);

Error in SO3Fun/discreteSample (line 23)
rot = SO3G(discretesample(d,n));

Error in coeff_fourier (line 279)
ori = discreteSample(odf,3000);
 
>> coeff_fourier
Warning: Imaginary part of complex valued SO3FunHarmonic is ignored.
In the following only the real part is plotted. 
> In SO3FunHarmonic/plotPDF (line 32)
In coeff_fourier (line 277) 
Error using histcounts>validateBinEdges
'BinEdges' must be a non-empty, real, numeric vector with no missing
values.

Error in histcounts (line 123)
        validateBinEdges(a);

Error in discretesample (line 108)
c = histcounts(rv, edges);

Error in SO3Fun/discreteSample (line 23)
rot = SO3G(discretesample(d,n));

Error in coeff_fourier (line 278)
ori = discreteSample(odf,3000);

\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

I'm still new with mtex and I hope have been able to clearly express my issues.

 I look forward to your answer. 

Many thanks. 

Moctar.

**Comment by num3RIK** [comment]
Hi Moctar,

thanks for your request.

First of all the warning is constructed since the ODF which is build by MTEX has complex valued function values instead of real valued as it may be expected. This has something to do with symmetric fourier coefficients. Moreover it is not necessary for them to be real valued to assure that the ODF is real valued.

I think something is going wrong by loading the data. Maybe by transfering them to MTEX syntax.

Unfortunately i am not able to reconstruct your error, since i do not have your data.
Therefore it would be nice if you could send me the data and youre complete script up to this point.
Then i will take a look on it and try to fix the problem.

Kind regards and thanks again for your request.

Erik Wünsche

PS: You can also send me the data privately by e-mail.

**Reply by moctarkaba** [reply]
Hi Erik,

Thanks for taking the time to reply to me.

As you mentioned, the issue is indeed the symmetry of these coefficients. I truly appreciate your offer, but I'm not authorized to share this data.

However, since the issue is more about the structure of the vector formed with these coefficients and not the data itself, I have sent you the structure of the code I have written for these ODF calculations via email. The email includes some data with the same structure (same sign of the values as the ones I'm working with, but with random values) along with all the necessary context.

Thanks again for your availability.
Best regards.

Moctar

**Comment by num3RIK** [comment, marked as answer]
Hi,
i just want to share the solution:
				
There exists lots of different definitions for the generalized spherical harmonics (Wigner-D functions), which distinguish by signs, complex conjugation, the order of the indices and the normalization factor. The definition which is used in MTEX can be found in the [documentation](https://mtex-toolbox.github.io/WignerFunctions.html).
Hence if you import the coefficients from another system to MTEX, you have to transform the them. Therefore you have to compute this transformation manually.

In case of VPSC you may use the Matlab-Code

> bandwidth = 4;
> fhat = zeros(deg2dim(bandwidth+1),1);
> for n = 0:bandwidth
>   ind = deg2dim(n)+1:deg2dim(n+1);
>   % transform to mtex coefficients
>   C_n = reshape(C(ind),2*n+1,2*n+1);
>   fhat(ind) = 1/sqrt(2*n+1) * (-1).^(min(-n:n,0)+min(-n:n,0)') .* C_n.';
> end
> odf = SO3FunHarmonic(fhat,cs)

I hope this works. ( Maybe try to compare the returned odf to that one you had in VPSC (The simplest way would be to evaluate both in the same orientations).

However I want to mention, that there is a much better way to load a odf into MTEX:
Usually VPSC does not calculate the harmonic coefficients. Usually VPSC works with the orientations and function values.
Also, it is very easy to load orientations and values into MTEX and then calculate the harmonic coefficients within MTEX.
If you go this way, you don't have to enter the coefficients all manually and you don't have to think about the definition (normalizaton, sign,...) of the Wigner-D functions.

Kind Regards
Erik

**Comment by moctarkaba** [comment]
hi Erik,

I'm truly sorry for not responding I was a bit busy (+ I took some days
off),
- first of all, thank you for your reply  and the script you send me
but, unfortunately it does not work the main issue remaining is the same
and plus there was some mistakes
in the data I have been given.

- Also I have been given a descretize orientation file which allow me to
keep going with my work.

- and for why I'm using VPSC here is attached a PDF for the abstract of my
work.

Best regards

Moctar.



Le mer. 14 août 2024 à 09:50, Erik Wünsche ***@***.***> a
écrit :
